python_modules/dagster/dagster/core/execution/context/compute.py

```diff
@@ -237,6 +246,48 @@ def get_tag(self, key: str) -> Optional[str]:
         """
         return self._step_execution_context.get_tag(key)
 
+    def has_events(self) -> bool:
+        return bool(self._events)
+
+    def retrieve_events(self) -> Iterator[DagsterEvent]:
+        while self._events:
+            yield self._events.pop(0)
```

Comment:
Why do we need to pop these?

Reply:
I'm not 100% following. Why not just return the list of events and have the calling code do `yield from`?

Reply:
If you take a look at [2], we're getting all the user-generated events that might happen between two different outputs being yielded so that we preserve the order. If we just returned the entire list of events, it wouldn't preserve order with things being yielded.

Reply:
I see. So it's a result of supporting both yield and log. Can we give the method a name that makes it clear that the operation is destructive? E.g. `dequeue_events`?

Also, not a huge deal because N is small, but I think the implementation might be unnecessarily N^2? If popping the first element of a list means shifting everything over a position. Could just assign `self._events` to the empty list and then return it?

Reply:
yea you're right that it's N^2. I like the proposed solution, will update with that

Reply:
`consume_events`?

Reply:
@alangenfeld had the idea of `report_events`